Client/mods/deathmatch/logic/CClientEntity.cpp

```diff
@@ -279,12 +279,10 @@ void CClientEntity::SetID(ElementID ID)
     }
 }
 
-CLuaArgument* CClientEntity::GetCustomData(const char* szName, bool bInheritData, bool* pbIsSynced)
+CLuaArgument* CClientEntity::GetCustomData(const SString& strName, bool bInheritData, bool* pbIsSynced)
```

Comment:
Yes, I know. But in this partucular cases `SString` was chosen for the reason that it was already used in almost all associative containers. By using lets say `std::string` you would get an additional string copy (see `SString(const std::string& strText`) constructor) and potential problems with key mathching. So the usage of std::string would exceed the boundaries of this PR.

Reply:
Why not use `std::string_view` all the way, and only in the end (when inserting) create the `SString`?
This way for the getter you can get away without creating a string at all.

Reply:
std::string_view uses 8 bytes.
Pointers and (I believe) references use 4 bytes.
So function use less registers. And the code can be optimized better.

Reply:
I think not allocating any memory until absolutely necessary far outweighs using 2 registers instead of 1.......

Reply:
Can you be more specific, please.
When we use 'getElementData' or 'setElementData', string is already created. 
Maybe, only 'CClientCClientEntity::SetCustomData' can be overloaded with 'std::string_view' key argument if we allow using data directly from network packets.
Where do you suggest use caching?

Reply:
The string is created within Lua, yes, so you can use a `string_view`, no need for `std::string` (Or `SString`).
Eventually when storing the key in the `std::map` the string has to be allocated.
But for getters `std::map` supports transparent lookup, so you can pass in `std::string_view` even if the key type is `std::string`.
(You just have to pass in `std::less<>` as the comparator instead of using the default one)

Reply:
Yeah. It seems `std::string_view` works without making a string copy for function arguments, but still makes a copy for output (this is expected). MTA has a very few usages of `std::string_view` in parser. So it wasn't expected, that this thing works.
getElementData can be optimized 100%.
Maybe `const char*` will be ever faster in this case.

Reply:
the argument parser correctly handles `std::string_view`.
`const char*` wont be faster, because it has to calculate the length, just use `string_view` that takes the length from Lua....
Why are you so against `string_view` I don't understand.

Reply:
Okay, I'll take into account your suggestions. std::string_view should fit in.

Reply:
> Why are you so against `string_view` I don't understand.

Im not totally against `string_view`. If you can pass simpler value, you should use it. Basic string compare function doesn't require string length. It's similar to removing unused data from function arguments.

Reply:
> > Why are you so against `string_view` I don't understand.
> 
> Im not totally against `string_view`. If you can pass simpler value, you should use it. Basic string compare function doesn't require string length. It's similar to removing unused data from function arguments.

`std::hash` function requires string len. That is `std::unordered_map` will have to construct a new string from the characters buffer(const char*) which is obviously not the best behaviour. `std::string_view` does not have this drawback.

Reply:
> `std::hash` function requires string len.

This is the answer. Thx

Reply:
> > > Why are you so against `string_view` I don't understand.
> > 
> > 
> > Im not totally against `string_view`. If you can pass simpler value, you should use it. Basic string compare function doesn't require string length. It's similar to removing unused data from function arguments.
> 
> `std::hash` struct requires string len. That is `std::unordered_map` will have to construct a new string from the characters buffer(const char*) which is obviously not the best behaviour. `std::string_view` does not have this drawback.

If you use std::map you can use take advantage of transparent lookup.
Sadly unordered_map doesn't support this.
Another way of doing this would be to use a fixed length string buffer as the key to avoid heap allocation (and by using an `unodered_map` we could increase cache locality)

Reply:
> > > > Why are you so against `string_view` I don't understand.
> > > 
> > > 
> > > Im not totally against `string_view`. If you can pass simpler value, you should use it. Basic string compare function doesn't require string length. It's similar to removing unused data from function arguments.
> > 
> > 
> > `std::hash` struct requires string len. That is `std::unordered_map` will have to construct a new string from the characters buffer(const char*) which is obviously not the best behaviour. `std::string_view` does not have this drawback.
> 
> If you use std::map you can use take advantage of transparent lookup. Sadly unordered_map doesn't support this. Another way of doing this would be to use a fixed length string buffer as the key to avoid heap allocation (and by using an `unodered_map` we could increase cache locality)

Heterogeneous lookups for unordered maps should be available since C++20. Regretfully we're still with C++17.

As you must be aware we are using both (unordered hash and tree-based maps) in MTA. Thus the solution must be selected according to the specifics of the both. Fixed length strings cannot be efficient here due to unpredictable length of such strings.

Reply:
> As you must be aware we are using both (unordered hash and tree-based maps) in MTA. Thus the solution must be selected according to the specifics of the both. Fixed length strings cannot be efficient here due to the unpredictable length of such strings.

See `MAX_CUSTOMDATA_NAME_LENGTH`.
So no name can be longer than 128 chars, so a 128 + 1 byte buffer would be enough.
A cache line is typically 64B, so keep that in mind. (But I think 128 + 1 for the char buffer, and uint8 for the length should be enough, this way the whole struct would be 132 bytes (I think, or 136 on 64b).
I'm unsure why you think fixed length strings would be inefficient. Tree based maps have extreme memory overhead for small values (std::string + CLuaArgument is barely a few bytes).
[Also, memory footprint can be lowered by merging my CLuaArgument refactor that uses `std::variant` (Thus the final size is quite a bit less)]

> Heterogeneous lookups for unordered maps should be available since C++20. Regretfully we're still with C++17

So we can upgrade to c++20 (For the server/client dm project at least).

Reply:
Also, we could experiment with custom hash functions (that generate less collisions for our use-case), if we decide to use a hashmap.

Reply:
> o

128 bytes is still a lot actually. On large servers such keys can eat up to 10 megabytes of the ram. You might say that 10 mb does not deserve attention, but on client side it may be significant (especially on high-load custom maps).

Personally I would prefer the StringName(as like in Godot Engine, or FName in UE) approach that implies just one buffer for each unique string. This way we only need a unique number that can be used as a key within the maps.

Reply:
That is an awesome idea, never thought of that, you're a genius!
So we could both save RAM, and CPU!
Just have to have a string interning system (That is global to all elements).
We also need to use some kind of reference counting (`shared_ptr` perhaps? Though the control block is allocated separately, that's kinda a waste of memory in our case) - To free unused strings.